File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(r, c):
            if (r < 0 or r >= len(grid) or
                c < 0 or c >= len(grid[r]) or
                grid[r][c] == "0"):
                return
            
            grid[r][c] = "0"
            dfs(r - 1, c)
            dfs(r + 1, c)
            dfs(r, c - 1)
            dfs(r, c + 1)

        count = 0
        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == "1":
                    dfs(r, c)
                    count += 1

        return count

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int island_count = 0;
        for (int row = 0; row < grid.length; ++row) {
            for (int col = 0; col < grid[0].length; ++col) {
                if (grid[row][col] == '1') {
                    dfs(row, col, grid);
                    island_count += 1;
                }
            }
        }
        
        return island_count;
    }
    
    private void dfs(int r, int c, char[][] matrix) {
        if (r < 0 || r == matrix.length ||
            c < 0 || c == matrix[0].length ||
            matrix[r][c] == '0') {
            return;
        }
        
        matrix[r][c] = '0';
        dfs(r - 1, c, matrix);
        dfs(r + 1, c, matrix);
        dfs(r, c - 1, matrix);
        dfs(r, c + 1, matrix);
    }
}